### PR TITLE
TEST: Drop nose for pytest

### LINKED ISSUE #864: Sprint: Pytest conversion
## Tasks

* Doctest modules
  * [x] Enable --doctest-modules in pytest
    * @effigies
* Failing modules
  * [x] [nibabel.streamlines.tests.test_array_sequence](https://github.com/nipy/nibabel/blob/pytest/nibabel/streamlines/tests/test_array_sequence.py)
    * @robbisg: #870 
  * [x] [nibabel.streamlines.tests.test_tck](https://github.com/nipy/nibabel/blob/pytest/nibabel/streamlines/tests/test_tck.py)
  * [x] [nibabel.streamlines.tests.test_tractogram](https://github.com/nipy/nibabel/blob/pytest/nibabel/streamlines/tests/test_tractogram.py)
    * @robbisg: #870 
  * [x] [nibabel.streamlines.tests.test_trk](https://github.com/nipy/nibabel/blob/pytest/nibabel/streamlines/tests/test_trk.py)
* Skipped modules (@djarecka: "I'll work on all of these")
  * [x] [nibabel.tests.test_scripts](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_scripts.py)
  * [x] [nibabel.tests.test_spaces](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_spaces.py)
  * [x] [nibabel.tests.test_spatialimages](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_spatialimages.py)
  * [x] [nibabel.tests.test_spm2analyze](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_spm2analyze.py)
  * [x] [nibabel.tests.test_spm99analyze](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_spm99analyze.py)
  * [x] [nibabel.tests.test_testing](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_testing.py)
  * [x] [nibabel.tests.test_tmpdirs](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_tmpdirs.py)
  * [x] [nibabel.tests.test_trackvis](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_trackvis.py)
  * [x] [nibabel.tests.test_tripwire](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_tripwire.py)
  * [x] [nibabel.tests.test_viewers](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_viewers.py)
  * [x] [nibabel.tests.test_volumeutils](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_volumeutils.py)
* Uncollected tests
  * [x] #879 @effigies
    * [ ] [nibabel/externals/tests/test_netcdf.py:48](https://github.com/nipy/nibabel/blob/pytest/nibabel/externals/tests/test_netcdf.py#L48): PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_read_write_files will be ignored
    * [ ] [nibabel/externals/tests/test_netcdf.py:89](https://github.com/nipy/nibabel/blob/pytest/nibabel/externals/tests/test_netcdf.py#L89): PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_read_write_sio will be ignored
    * [ ] [nibabel/externals/tests/test_netcdf.py:140](https://github.com/nipy/nibabel/blob/pytest/nibabel/externals/tests/test_netcdf.py#L140): PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_write_invalid_dtype will be ignored
  * [x] [nibabel/tests/test_image_api.py:715](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_image_api.py#L715): PytestCollectionWarning: cannot collect test class 'TestMinc2API' because it has a \_\_init\_\_ constructor
  * [x] [nibabel/tests/test_scripts.py:104](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_scripts.py#L104): PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_nib_ls will be ignored
  * [x] [nibabel/tests/test_scripts.py:187](https://github.com/nipy/nibabel/blob/pytest/nibabel/tests/test_scripts.py#L187): PytestCollectionWarning: yield tests were removed in pytest 4.0 - test_nib_diff will be ignored
    * @djarecka 
* Untouched submodules
  * [ ] nibabel.benchmarks
  * [x] nibabel.cifti2.tests
    * @robbisg
  * [x] nibabel.cmdline.tests
  * [x] nibabel.externals.tests
  * [x] nibabel.freesurfer.tests
    * @orduek 
  * [x] nibabel.gifti.tests
    * @anibalsolon
  * [x] nibabel.nicom.tests
    * #884 @effigies
* Refactors
  * [x] Skip base test cases like [_TestWrapStructBase](https://github.com/nipy/nibabel/blob/f908ca4b5ed690a953c89af45656df558b733514/nibabel/tests/test_wrapstruct.py#L109).
    * Possible idea: use `setUp()` to test whether the class name starts with `_`, and `pytest.skip()`.
    * @effigies: #871
  * [ ] Use `pytest.deprecated_call()` and `pytest.warns()` instead of `clear_and_catch_warning`, when sensible. See https://github.com/nipy/nibabel/pull/865#discussion_r370957350 and https://github.com/nipy/nibabel/pull/865#discussion_r370957571.
    * [ ] Related: [assert_warn_len_equal](https://github.com/nipy/nibabel/blob/ddf26830eebdb11470009c3a04047377f29a4732/nibabel/tests/test_testing.py#L52-L84) is kind of a mess, and made redundant with going Python 3+.

Some principles:
* Try to avoid numpy test methods and decorators when a vanilla [unittest](https://docs.python.org/3/library/unittest.html)/[pytest](https://docs.pytest.org/) construct exists. NOTABLE EXCEPTIONS: Tools for comparing arrays and floating points, such as `assert_almost_equal`.
* Clear is better than brief.
* ...

Original post follows:

-----

## Summary

`nose` is finally going to start failing on Python 3.9 due to some standard library API changes.

Python 3.9.0 is due to be released in October, but we might as well get out ahead of it. We don't need to rush, so I'm going to target the nibabel 3.1.0 release.

I put out a [call on Twitter](https://twitter.com/effigies/status/1220097622276562945) to see if anybody wanted to sprint on this, and got responses from @arokem, @chrisgorgo, @orduek, @robbisg and @anibalsolon (thanks to all for having the same handle on GitHub and Twitter).

This thread exists to coordinate the sprint.

## Status

@djarecka and @chrisgorgo started on conversion in November:

* [x] #835
* [x] #838 
* [x] #839
* [x] #840 

There have been some changes in the tests since #840 was started, so it requires merging/rebasing.

## Plans

To facilitate the transition, I've started the [`pytest`](https://github.com/nipy/nibabel/tree/pytest) branch, which is #840's commits, rebased on master.

The first thing to do is to identify all remaining tasks. Off the top of my head we need to:

1) Eliminate the use of `nose.tools`.
2) Move any nose test case classes to vanilla `unittest` features, which pytest should be able to handle transparently.
3) Verify coverage doesn't change/increases.
4) Run doctests in the documentation with pytest.

## Logistics

Let's set at least one definite window that people want to be actively working on this and where I can commit to being more responsive to keep things from stalling out. We don't all need to align, but it would probably be best for me if there isn't a separate window for each participant.

Would it be useful to attempt to do this (semi-)synchronously, and use a chat room? Or would people rather go through the usual process of creating issues and pull requests?

### PR DESCRIPTION
**Pull request by effigies**
This PR exists to make all changes related to #864 easy to view, and makes CI status more visible.

Closes #864.
Replaces #840.

### CODE REVIEW

File: nibabel/tests/test_proxy_api.py [hunk 1]
```diff
@@ -105,24 +102,26 @@ def validate_shape(self, pmaker, params):
         prox, fio, hdr = pmaker()
         assert_array_equal(prox.shape, params['shape'])
         # Read only
-        assert_raises(AttributeError, setattr, prox, 'shape', params['shape'])
+        with pytest.raises(AttributeError):
+            setattr(prox, 'shape', params['shape'])
```
Comment (effigies) [inline nibabel/tests/test_proxy_api.py]: Should probably just `grep setattr` to find all of these read-only checks.

File: nibabel/tests/test_quaternions.py [hunk 2]
```diff
@@ -55,129 +53,135 @@ def test_fillpos():
...
+    assert nq.norm(qi) == 1
+    assert nq.isunit(qi)
     qi[1] = 0.2
-    yield assert_true, not nq.isunit(qi)
+    assert not nq.isunit(qi)
 
 
-@slow
-def test_mult():
+@dec.slow
```
Comment (effigies) [inline nibabel/tests/test_quaternions.py]: Should see if the equivalent can be done with pure unittest/pytest constructs. My impression is that numpy didn't really intend a lot of its testing infrastructure to be a public API.

File: nibabel/tests/test_recoder.py [hunk 3]
```diff
@@ -12,50 +12,75 @@
...
     # simplest case, no aliases
     codes = ((1,), (2,))
     rc = Recoder(codes)
-    yield assert_equal, rc.code[1], 1
-    yield assert_equal, rc.code[2], 2
-    yield assert_raises, KeyError, rc.code.__getitem__, 3
+    assert rc.code[1] == 1
+    assert rc.code[2] == 2
+    with pytest.raises(KeyError):
+        rc.code.__getitem__(3)
```
Comment (effigies) [inline nibabel/tests/test_recoder.py]: `__getitem__` calls should also be grepped for.

```suggestion
        rc.code[3]
```

File: nibabel/tests/test_recoder.py [hunk 4]
```diff
@@ -12,50 +12,75 @@
...
+
+def test_recoder_2():
     # with explicit name for code
+    codes = ((1,), (2,))
     rc = Recoder(codes, ['code1'])
-    yield assert_raises, AttributeError, rc.__getattribute__, 'code'
-    yield assert_equal, rc.code1[1], 1
-    yield assert_equal, rc.code1[2], 2
+    with pytest.raises(AttributeError):
+        rc.__getattribute__('code')
```
Comment (effigies) [inline nibabel/tests/test_recoder.py]: And `__getattribute__` and `getattr`:

```suggestion
        rc.code
```

File: nibabel/tests/test_recoder.py [hunk 5]
```diff
@@ -12,50 +12,75 @@
...
+    with pytest.raises(KeyError):
+        rc.code.__getitem__(3)
+    assert rc.code['one'] == 1
+    assert rc.code['two'] == 2
+    with pytest.raises(KeyError):
+        rc.code.__getitem__('three')
+    with pytest.raises(AttributeError):
+        rc.__getattribute__('label')
+
+def test_recoder_3():
```
Comment (effigies) [inline nibabel/tests/test_recoder.py]: Doubled up. We might be masking the first one of these...

File: nibabel/tests/test_scaling.py [hunk 6]
```diff
@@ -129,54 +131,63 @@ def test_a2f_nan2zero():
...
+    slope, inter, mn, mx = _calculate_scale(arr, out_dtype, True)
+    array_to_file(arr, bio, out_type, 0, inter, slope, mn, mx)
+    bio.seek(0)
+    arr2 = array_from_file(arr.shape, out_dtype, bio)
+    arr3 = apply_read_scaling(arr2, slope, inter)
+    # Max rounding error for integer type
+    max_miss = slope / 2.
+    assert np.all(np.abs(arr - arr3) <= max_miss)
+    bio.truncate(0)
+    bio.seek(0)
```
Comment (effigies) [inline nibabel/tests/test_scaling.py]: We can drop these last two lines, since we're not reusing `bio`.

File: nibabel/tests/test_scaling.py [hunk 7]
```diff
@@ -129,54 +131,63 @@ def test_a2f_nan2zero():
...
+
+
+@pytest.mark.parametrize("category0, category1",[
```
Comment (effigies) [inline nibabel/tests/test_scaling.py]: Should think about if there are more informative names here.

File: nibabel/tests/test_wrapstruct.py [hunk 8]
```diff
@@ -120,15 +119,17 @@ def get_bad_bb(self):
         return None
 
     def test_general_init(self):
+        if not self.header_class:
+            pytest.skip()
```
Comment (effigies) [inline nibabel/tests/test_wrapstruct.py]: I guess `nose` was skipping `_TestWrapStructBase`, but `pytest` isn't... Possibly we can do something with `setUp()` to ensure that the base class doesn't run but the subclasses do.

Another option would be to provide an absolute minimal `header_class` to test on.

**Comment by djarecka** [comment]
@effigies - thanks for reviewing, I'll address the comments and change the code before Tue

**Comment by effigies** [comment]
@djarecka I just wanted to establish a baseline status. I'm okay with using this as a starting point and divvying up jobs on Tuesday.

Of course, feel free to push on anything you like before then.

**Comment by robbisg** [comment]
Maybe we can already sketch and/or discuss how to divide jobs in these days.

File: .travis.yml [hunk 9]
```diff
@@ -124,12 +128,72 @@ script:
...
+                -I test_pkg_info \
+                -I test_processing \
+                -I test_proxy_api \
+                -I test_quaternions \
+                -I test_recoder \
+                -I test_remmovalschedule \
+                -I test_round_trip \
+                -I test_rstutils \
+                -I test_scaling \
+                -I test_wrapstruct
```
Comment (effigies) [inline .travis.yml]: When a file is covered under pytest,  add it to this list.

**Comment by effigies** [comment]
## Quick tutorial on merging/rebasing the base branch

If you don't have the `nipy` repository as the upstream remote:

```
git remote add upstream git@github.com:nipy/nibabel.git
```

Fetch the latest state:

```
git fetch upstream
```

If you want to merge (easy):

```
git merge upstream/pytest
```

If you want to rebase (can be tricky):

```
git rebase -i upstream/pytest
```

But if you're not comfortable with that, it's probably not worth the effort to learn today.

File: nibabel/tests/test_nifti1.py [hunk 10]
```diff
@@ -61,7 +60,43 @@
...
+        ((-np.inf, -np.inf), HDE, (None, None), (-np.inf, -np.inf)),
+        # Good slope and bad inter generates error for get_slope_inter
+        ((2, None), HDE, HDE, (2, np.nan)),
+        ((2, np.nan), HDE, HDE, (2, np.nan)),
+        ((2, np.inf), HDE, HDE, (2, np.inf)),
+        ((2, -np.inf), HDE, HDE, (2, -np.inf)),
+        # Good slope and inter - you guessed it
+        ((2, 0), None, (2, 0), (2, 0)),
+        ((2, 1), None, (2, 1), (2, 1))
+    ]
```
[the remark this answers is not in the document]
Comment (djarecka) [reply nibabel/tests/test_nifti1.py]: not sure if I understand your comment, but I know you were working on this test anyway

File: nibabel/tests/test_nifti1.py [hunk 11]
```diff
@@ -285,54 +289,52 @@ def test_freesurfer_large_vector_hack(self):
...
-        assert_raises(HeaderDataError, hdr.set_data_shape, (too_big, 1, 2))
-        assert_raises(HeaderDataError, hdr.set_data_shape, (too_big, 2, 1))
-        assert_raises(HeaderDataError, hdr.set_data_shape, (1, too_big))
-        assert_raises(HeaderDataError, hdr.set_data_shape, (1, too_big, 1))
-        assert_raises(HeaderDataError, hdr.set_data_shape, (1, 1, too_big))
-        assert_raises(HeaderDataError, hdr.set_data_shape, (1, 1, 1, too_big))
+        for args in [(too_big,), (too_big, 1), (too_big, 1, 2), (too_big, 2, 1),
+                     (1, too_big), (1, too_big, 1), (1, 1, too_big), (1, 1, 1, too_big)]:
+            with pytest.raises(HeaderDataError):
+                hdr.set_data_shape(args)
```
[the remark this answers is not in the document]
Comment (djarecka) [reply nibabel/tests/test_nifti1.py]: this is in many places: i even didn't think about this one line syntax, when I was changing the tests...

File: nibabel/tests/test_openers.py [hunk 12]
```diff
@@ -158,22 +160,26 @@ def test_Opener_gzip_type():
             with patch_indexed_gzip(igzip_present):
                 assert isinstance(Opener(fname, **kwargs).fobj, expected)
 
+@pytest.fixture(scope="class")
```
[the remark this answers is not in the document]
Comment (djarecka) [reply nibabel/tests/test_openers.py]: not sure

File: nibabel/tests/test_parrec_data.py [hunk 13]
```diff
@@ -63,7 +60,8 @@ def test_fieldmap():
     fieldmap_nii = pjoin(BALLS, 'NIFTI', 'fieldmap.nii.gz')
     load(fieldmap_par)
     top_load(fieldmap_nii)
-    raise SkipTest('Fieldmap remains puzzling')
+    # TODO dj: i believe this shouldn't be here
+    #raise pytest.skip('Fieldmap remains puzzling')
```
[the remark this answers is not in the document]
Comment (djarecka) [reply nibabel/tests/test_parrec_data.py]: ok, reverted

**Comment by djarecka** [comment]
@effigies - i didn't notice many of your comments before, were hidden by github and didn't realize it.
I resolved some of them (will commit and push shortly) and comment on rest of them.

File: .travis.yml [hunk 14]
```diff
@@ -26,17 +26,16 @@ python:
...
+        - CHECK_TYPE="nosetests"
     # Basic dependencies only
     - python: 3.5
       env:
         - DEPENDS="-r requirements.txt"
-    # Clean install
-    - python: 3.5
-      env:
-        - DEPENDS=""
-        - CHECK_TYPE=skiptests
```
Comment (effigies) [inline .travis.yml]: I regret this... The `DEPENDS=""` is doing the work here and is still needed to test that `nibabel` will work when no dependencies are pre-installed.

**Comment by djarecka** [comment]
@effigies - I've realized how bad my git history is... perhaps I can try to remove some of the commits, but would destroy the history...

**Comment by effigies** [comment]
Eh, not really worth the effort at this point, IMO. It's just some duplicated commits.

**Comment by effigies** [comment]
Hi all, this is ready, as far as I'm concerned. There are further things that we can do, but the task of removing the dependency on `nose` is done, so my inclination is to get it merged, and let any further work be done against `master`, rather than `pytest`.

There's a slightly weird issue where the `def <func>():` lines aren't getting covered, and that's probably a result of using `--pyargs` to test the installed version rather than directly testing the source. I'm not sure fiddling with this is a great use of time.

What do y'all think? Anything you see as hold-ups?

**Comment by effigies** [comment]
Okay. I'm doing it. Thanks all for sprinting on this with me!

**Pull request merged**